### PR TITLE
Re-write text wrapping logic

### LINKED ISSUE #15014: Wrapping of text adds leading newline character if first word is long
### Bug report

**Bug summary**

Setting wrap=True on a text object having a first word exceeding the line width returns a leading newline character.

**Code for reproduction**

```python
import matplotlib.pyplot as plt

fig = plt.figure(figsize=(6, 4))
plt.axis([0, 10, 0, 10])
text = plt.text(9.5, 8, 'Alonglineoftexttowrap', ha='left', va='top', wrap=True)
plt.text(4, 8, 'Some text as position reference', ha='left', va='top', wrap=True)
plt.show()

print(repr(text._get_wrapped_text()))
```

**Actual outcome**

```
'\nAlonglineoftexttowrap'
```
wrap_output

**Expected outcome**

Assuming the leading newline character is unnecessary, only expected 'Alonglineoftexttowrap'.

**Matplotlib version**

  * Operating system: OS X
  * Matplotlib version: 3.0.3
  * Matplotlib backend (`print(matplotlib.get_backend())`): MacOSX
  * Python version: 3.7.3
  * Jupyter version (if applicable):
  * Other libraries:

### PR DESCRIPTION
**Pull request by dstansby**
I did not really understand the previous logic, and I think it got some things wrong (see below), so I've re-written it. I hope this is more understandable... in particular

- Doesn't include spurious newlines at the beginning of long single word lines (fixes #15014)
- Doesn't include extra newlines at the end of wrapped text
- Hopefully the code is more readable

### CODE REVIEW

**Comment by timhoffm** [comment]
Is there something special in the wrapping algorithm that prevents us from using the standardlib https://docs.python.org/3.7/library/textwrap.html?

**Comment by jklymak** [comment]
Its specified in window extents, not characters...

**Comment by dstansby** [comment]
I guess because python std library wraps to a certain number of characters, whereas we want to wrap a rendered on screen width, which are not necessarily equivalent (unless using a fixed width font)

**Comment by jklymak** [comment]
This looks good, but can you add tests for the cases above?

**Comment by jklymak** [comment]
I guess a possible optimization would be to figure out how wide "M" is and use that as the character width rather than checking the width of each line after adding each word.  That would be a lot faster, but would of course occasionally wrap too soon.

**Comment by anntzer** [comment]
> That would be a lot faster, but would of course occasionally wrap too soon.

I would not bother unless someone has an application where text wrapping is the bottleneck (and if anything we could fix that by caching the result of the length measurements after each word addition -- we're going to need to position the glyphs to actually render the text, in any case).

**Comment by tacaswell** [comment]
It looks like we don't have any other tests of wrapped text?

**Comment by timhoffm** [comment]
This is not working. I took the liberty to push a test that displays the failure.

master:
grafik

this:
grafik

**Comment by dstansby** [comment]
Thanks for spotting that, it was missing a couple of breaks.

**Comment by PatrickFeiring** [comment]
Whilst at it, are there any plans on supporting wrapping at something other than the figure boundaries, which is the only thing that is supported at the moment if I am not mistaken? Taking a user-specified width would make it possible to place boxes of text anywhere in the figure.

From what I can tell most of the logic is already in place, only the line in _get_wrapped_text returning the desired line length needs to change, 
```python
# Build the line incrementally, for a more accurate measure of length
line_width = self._get_wrap_line_width()
```
of course, in addition to the ability to specify the width somewhere.

**Comment by timhoffm** [comment]
> supporting wrapping at something other than the figure boundaries

Would be a nice addition and probably not too hard. But that's something for a separate feature request.

File: lib/matplotlib/text.py [hunk 1]
```diff
@@ -632,27 +632,40 @@ def _get_wrapped_text(self):
...
+
+        # Now wrap each individual unwrapped line
+        for unwrapped_line in unwrapped_lines:
+
+            sub_words = unwrapped_line.split(' ')
+            # Remove items from sub_words as we go, so stop when empty
+            while len(sub_words) > 0:
+                if len(sub_words) == 1:
+                    # Only one word, so just add it to the end
+                    wrapped_str += sub_words.pop(0)
```
Comment (anntzer) [inline lib/matplotlib/text.py]: What about having wrapped_lines as a list of strings that you "\n".join() at the very end (or even just return the list of lines and change the call site?  I think right now one difficulty to follow the flow of the algorithm is that the place where the linebreaks are introduced is a bit hidden (it's a `+ '\n'` that hides at the end of a line).

**Pull request merged**